update: dynamic library tutorial (#4556)
JetBrains/kotlin-web-site: docs/topics/native/native-dynamic-libraries.md

   the library:

   ```bash
   ./gradlew linkDebugSharedNative
   ```

The build generates the library into the `build/bin/native/debugShared` directory with the following files:

* macOS `libnative_api.h` and `libnative.dylib`
* Linux: `libnative_api.h` and `libnative.so`
* Windows: `libnative_api.h`, `libnative.def`, and `libnative.dll`

> You can also use the `linkNative` Gradle task to generate both `debug` and `release` variants of the library. 
> 
{style="tip"}

The Kotlin/Native compiler uses the same rules to generate the `.h` file for all platforms. Let's check out the C API of
the Kotlin library.

## Generated header file

Let's examine how Kotlin/Native declarations are mapped into C functions.

In the `build/bin/native/debugShared` directory, open the `libnative_api.h` header file.
The very first part contains the standard C/C++ header and footer:

```c
#ifndef KONAN_LIBNATIVE_H
#define KONAN_LIBNATIVE_H
#ifdef __cplusplus
extern "C" {
#endif

/// The rest of the generated code

#ifdef __cplusplus
}  /* extern "C" */
#endif
#endif  /* KONAN_LIBNATIVE_H */
```

> After the rituals in the `libnative_api.h`, there is a block with the common type definitions:

Author: Yeah, good idea!

|------------------------|-----------------------------------------------|
| `libnative_KBoolean`   | `bool` or `_Bool`                             |
| `libnative_KChar`      | `unsigned short`                              |
| `libnative_KByte`      | `signed char`                                 |
| `libnative_KShort`     | `short`                                       |
| `libnative_KInt`       | `int`                                         |
| `libnative_KLong`      | `long long`                                   |
| `libnative_KUByte`     | `unsigned char`                               |
| `libnative_KUShort`    | `unsigned short`                              |
| `libnative_KUInt`      | `unsigned int`                                |
| `libnative_KULong`     | `unsigned long long`                          |
| `libnative_KFloat`     | `float`                                       |
| `libnative_KDouble`    | `double`                                      |
| `libnative_KVector128` | `float __attribute__ ((__vector_size__ (16))` |
| `libnative_KNativePtr` | `void*`                                       |

> The definition part of the `libnative_api.h` file shows how Kotlin primitive types map into C primitive types.

Author: I kept the beginning because here the most part is automatically generated definitions, which we don't describe in particular

          libnative_kref_example_Clazz (*Clazz)();
          libnative_KULong (*memberFunction)(libnative_kref_example_Clazz thiz, libnative_KInt p);
        } Clazz;
        const char* (*get_globalString)();
        void (*forFloats)(libnative_KFloat f, libnative_KDouble d);
        void (*forIntegers)(libnative_KByte b, libnative_KShort s, libnative_KUInt i, libnative_KLong l);
        const char* (*strings)(const char* str);
      } example;
    } root;
  } kotlin;
} libnative_ExportedSymbols;
```

The code uses anonymous structure declarations. Here, `struct { ... } foo` declares a field in the outer struct of that
anonymous structure type, the type with no name.

> C doesn't support objects either, so function pointers are used to mimic object semantics. A function pointer is declared

Author: I kept 'either' because it's a reference that it doesn't support namespaces from earlier

as `RETURN_TYPE (* FIELD_NAME)(PARAMETERS)`.

There is a `libnative_kref_example_Clazz` field that represents the `Clazz` from Kotlin. The `libnative_KULong` is
accessible with the `memberFunction` field. The only difference is that the `memberFunction` accepts a `thiz` reference as
the first parameter. The C language doesn't support objects, which is why `thiz` pointer should be passed explicitly.

There is a constructor in the `Clazz` field (aka `libnative_kref_example_Clazz_Clazz`), which is the constructor function
to create an instance of the `Clazz`.

> Kotlin `object Object` is accessible as `libnative_kref_example_Object`. There's the `_instance` function to get the only

Author: no, it's correct, direct from the code block above

instance of the object.

Properties are translated into functions. The `get_` and `set_` prefixes name the getter and the setter functions,
respectively. For example, the read-only property `globalString` from Kotlin is turned into a `get_globalString`
function in C.

Global functions `forFloats`, `forIntegers`, and `strings` are turned into functions pointers in the `libnative_kref_example`
anonymous struct.

### Entry point

> You can see how the API is created. To start with, you need to initialize the `libnative_ExportedSymbols` structure.

Reviewer: Hmm 🤔 
This section is a bit difficult to follow in my opinion. Lots of things are happening at the same time - The user can see how the API created - then they have to initialize something, and then they need to look at the final part of the `libnative.api.h`.
 
Could it be cleaned up by saying something like: 
You can initialize the API using the `libnative_ExportedSymbols` structure as seen in the the final part of the `libnative_api.h`:

Or:

You can see how the API is initialized using the `libnative_ExportedSymbols` structure in the final part of the `libnative_api.h`:

Author: These first two sentences just connect all the previous header file description with this last part. But I can see how it maybe confusing, so I created a new version

Reviewer: Nice - it reads much better now thank you! 👍

Let's take a look at the final part of the `libnative_api.h` for this:

```c
extern libnative_ExportedSymbols* libnative_symbols(void);
```

The function `libnative_symbols` allows you to open the way from the native code to the Kotlin/Native library. This is
the entry point you'll use. The library name is used as a prefix for the function name.

> > Hosting the returned `libnative_ExportedSymbols*` pointer per thread might be necessary.

Reviewer: ```suggestion
> It might be necessary to host the returned `libnative_ExportedSymbols*` pointer per thread.
```

}
```

## Compile and run the project

### On macOS

To compile the C code and link it with the dynamic library, navigate to the library directory and run the following command:

```bash
clang main.c libnative.dylib
```

The compiler generates an executable called `a.out`. Run it to see the Kotlin code executed from the C library.

### On Linux

To compile the C code and link it with the dynamic library, navigate to the library directory and run the following command:

```bash
gcc main.c libnative.so
```

> The compiler generates an executable called `a.out`. Run it to see the Kotlin code executed from the C library. On

Reviewer: ```suggestion
The compiler generates an executable called `a.out`. Run it to execute the Kotlin code from the C library. On
```

Linux, you need to include `.` into the `LD_LIBRARY_PATH` to let the application know to load the `libnative.so`
library from the current folder.

### On Windows

First, you'll need to install a Microsoft Visual C++ compiler that supports the x64_64 target.

The easiest way to do this is to install Microsoft Visual Studio on a Windows machine. During installation,

> select components necessary to work with C++, for example, **Desktop development with C++**.

Reviewer: ```suggestion
select the necessary components for C++, for example, **Desktop development with C++**.
```

> On Windows, dynamic libraries are included either via a generated static library wrapper or with manual code,

Reviewer: ```suggestion
On Windows, you can include dynamic libraries either by generating a static library wrapper or manually handling DLL loading with
```

Reviewer: (Suggestion) - without latin and more active voice

> which deals with the [LoadLibrary](https://learn.microsoft.com/en-gb/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibrarya)

Reviewer: ```suggestion
[LoadLibrary](https://learn.microsoft.com/en-gb/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibrarya)
```
